src/lib/FlightTasks/tasks/ManualAltitudeSmoothVel/FlightTaskManualAltitudeSmoothVel.cpp

```diff
@@ -0,0 +1,167 @@
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file FlightManualAltitude.cpp
+ */
+
+#include "FlightTaskManualAltitudeSmoothVel.hpp"
+
+#include <float.h>
+
+using namespace matrix;
+
+bool FlightTaskManualAltitudeSmoothVel::activate()
+{
+	bool ret = FlightTaskManualAltitude::activate();
+
+	_reset();
+
+	return ret;
+}
+
+void FlightTaskManualAltitudeSmoothVel::reActivate()
+{
+	// The task is reacivated while the vehicle is on the ground. To detect takeoff in mc_pos_control_main properly
+	// using the generated jerk, reset the z derivatives to zero
+	_reset(true);
+}
+
+void FlightTaskManualAltitudeSmoothVel::_reset(bool force_vz_zero)
+{
+	// Set the z derivatives to zero
+	if (force_vz_zero) {
+		_smoothing.reset(0.f, 0.f, _position(2));
+
+	} else {
+		// TODO: get current accel
+		_smoothing.reset(0.f, _velocity(2), _position(2));
+	}
+
+	// Always start unlocked
+	_position_lock_z_active = false;
+	_position_setpoint_z_locked = NAN;
+}
+
+void FlightTaskManualAltitudeSmoothVel::_checkEkfResetCounters()
+{
+	if (_sub_vehicle_local_position->get().z_reset_counter != _reset_counters.z) {
+		_smoothing.setCurrentPosition(_position(2));
+		_reset_counters.z = _sub_vehicle_local_position->get().z_reset_counter;
+	}
+
+	if (_sub_vehicle_local_position->get().vz_reset_counter != _reset_counters.vz) {
+		_smoothing.setCurrentVelocity(_velocity(2));
+		_reset_counters.vz = _sub_vehicle_local_position->get().vz_reset_counter;
+	}
+}
+
+void FlightTaskManualAltitudeSmoothVel::_updateSetpoints()
+{
+	/* Get yaw setpont, un-smoothed position setpoints.*/
+	FlightTaskManualAltitude::_updateSetpoints();
+
+	/* Update constraints */
+	if (_velocity_setpoint(2) < 0.f) { // up
+		_smoothing.setMaxAccel(_param_mpc_acc_up_max.get());
+		_smoothing.setMaxVel(_constraints.speed_up);
+
+	} else { // down
+		_smoothing.setMaxAccel(_param_mpc_acc_down_max.get());
+		_smoothing.setMaxVel(_constraints.speed_down);
+	}
+
+	float jerk = _param_mpc_jerk_max.get();
```

Comment:
- does it make sense to use the same jerk limit as for xy?
- you overwrite `jerk` below unconditionally

Reply:
- We could have a different parameter for jerk Z but I think most people will leave it as is. I never had the feeling that I needed to adjust it until now.

- Right, I'll keep that in mind for the next iteration